candlepin/candlepin
# Reconnect to Artemis broker when connection goes down

* Created an ActiveMQConnectionMonitor that listens for connection drops
  and will run a connection monitor and notify listeners when the connection
  is re-established.
* EventSource will now listen for connection drops
  and will re-establish connections to Artemis when
  the connection is reported by the monitor as being
  up.
* Candlepin will now enter suspend mode when the ActiveMQ (Artemis) broker
  is down.
* Suspend mode transitioner now based on reasons since there could be multiple
  reasons reported as to why candlepin is in suspend mode. Only the initial
  reason will be shown in the /status response (API V1).
* EventSink will no longer attempt to send/rollback when no session exists
  which supresses the ActiveMQ rollback exceptions when no messages are sent
  during a particular request.
* Removed the use of ThreadLocal in the EventSink for Sessions and Producers
  - A single instance of the EventSink is now created per request and has a
    single ClientSession and Producer.


# Testing

1. Deploy candlepin with an external broker plus qpid
```bash
$ bin/deploy -gtaqb
```

2. Tail the logs in a separate console.

3. Shut down artemis
```bash
$ sudo systemctl stop artemis
```

4. Check the candlepin status and it should have entered suspend mode.
```bash
$ curl -k https://localhost:8443/candlepin/status
```

5. Try registering a consumer. It should fail because of suspend mode.
```bash
curl -k -u admin:admin -X POST -H "Content-Type: application/json" -d '{"type":{"label":"system"}, "name":"my_test_system", "facts":{}, "installedProducts":[], "contentTags":[]}' https://localhost:8443/candlepin/consumers?owner=admin
```

6. Restart Artemis, wait for candlepin to come out of suspend mode and attempt to register a consumer again. This time it should succeed.
```bash
$ sudo systemctl start artemis
```

7. Shut down qpidd and wait for suspend mode to be entered. Then shut down artemis. The logs should show that candlepin is in suspend mode for 2 reasons: qpidd and artemis are down. Candlepin status will only reflect the last state causing suspend mode (didn't want to change the API at this point).
```bash
$ sudo systemctl stop qpidd
$ sudo systemctl stop artemis
$ curl -k https://localhost:8443/candlepin/status
```

8. Start qpid and candlepin will remain in suspend mode.
```bash
$ sudo systemctl start qpidd
$ curl -k https://localhost:8443/candlepin/status
```

9. Start artemis and candlepin will return to Normal mode.
```bash
$ sudo systemctl start artemis
$ curl -k https://localhost:8443/candlepin/status
```

10. Register another system to ensure that all connections have been re-established.
```bash
curl -k -u admin:admin -X POST -H "Content-Type: application/json" -d '{"type":{"label":"system"}, "name":"my_test_system_2", "facts":{}, "installedProducts":[], "contentTags":[]}' https://localhost:8443/candlepin/consumers?owner=admin
```

NOTE: You can check the artemis console to make sure that messages are flowing.
http://localhost:8161/console

### Additional Testing
When candlepin is started without qpid and/or artemis started, candlepin should start Ok and should be in suspend mode.

Check it with by curling the candlepin status.
